### ISSUE TITLE
Airlock Processor Unable to Start

### ISSUE THREAD

**Issue by martinpeck** [microsoft/AzureTRE#3664]
We observed that a working/functioning deploy of the AzureTRE developed an issue where by airlock functionality stopped working. 

**Symptoms**

- The primary symptom that we observed was that storage containers for new import requests were not created. 
- In this state, messages were left unprocessed on `airlock-status-changed` service bus. 
- Additionally, in the azure portal, we saw an error in the portal when viewing the airlock processor function app that suggested that an SSL error had occurred. Also, the portal reported `error` for the function's runtime.

While this looked, at first, like an SSL issue further investigation showed the the airlock function was attempting to make outbound network calls during startup that were being blocked by the AzureTRE's firewall. Inspecting the firewall logs we discovered a number of DENY logs related to the function app runtime host attempting to connect to the following two endpoints:

- `functionscdn.azureedge.net`
- `func-airlock-processor-MYTRE.azurewebsites.net`

**Work Around**

Using the IP address of the function app opening up the firewall to allow outbound traffic has allowed the airlock function app to start up. The illustration below is clearly less constrained than it should be, but this patch to the firewall allows the function app to work. **Use with caution.**

![image](https://github.com/microsoft/AzureTRE/assets/1524451/6273651f-4453-4693-9ddd-1fc90681bdb1)

**Longer Term Fix**

I suggest the longer term fix is to ensure that `functionscdn.azureedge.net` is added to the allow-list on the firewall so that traffic from the function app is not blocked I would not suggest we add `func-airlock-processor-MYTRE.azurewebsites.net` until we're sure we understand why the function app is trying to call itself.

**Root Cause Analysis**

I believe the airlock function app was initially working because the function is deployed BEFORE the firewall is deployed. So, the function app was able to make this call successfully. Then, for reasons I'm unclear about, the function app was redeployed. This might have been a scale-out event, or might have been as part of some action performed by another admin, but at any rate the function app was redeployed and attempted to make this network call on startup. This failed as the firewall now blocked the call.

The version of the Azure Functions runtime that we're using is 3.20.

In the source code for 3.20 we can see that the URL `https://functionscdn.azureedge.net/public` is defined here
https://github.com/Azure/azure-functions-host/blob/v3.20.0/src/WebJobs.Script/ScriptConstants.cs#L175

This url is used in the `ExtensionBundleManager`:
https://github.com/Azure/azure-functions-host/blob/v3.20.0/src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs#L37

This is then used to check for, and download, bundle extensions:
https://github.com/Azure/azure-functions-host/blob/v3.20.0/src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs#L173

https://github.com/Azure/azure-functions-host/blob/v3.20.0/src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs#L231

And, for context, the function's `host.json` specifies a bundle extension...

``` json
  "extensionBundle": {
    "id": "Microsoft.Azure.Functions.ExtensionBundle",
    "version": "[3.3.0, 4.0.0)"
  }
```

So, I believe that this is being called when the function is first deployed, and the bundle extension is being downloaded at this point. With the firewall in place it's not possible to download this bundle extension.

The bundle extension is what allows non-dotnet languages to register triggers:
https://github.com/Azure/azure-functions-extension-bundles/tree/v3.x

**Comment by marrobi** [comment]
This seems to be the same issue - https://github.com/Azure/azure-functions-host/issues/8168. Agree we should add the rule to the firewall for the `AirlockProcessorSubnet` .

**Comment by martinpeck** [comment]
Yep. There are quite a few examples of this once you know what to look for.
Most refer to this becoming a problem with v4 runtime, but the code paths are pretty much the same for the v3 runtime.

I suggest we include a change to update the runtime at the same time.

**Comment by martinpeck** [comment]
I would assume so, yes

**Comment by bwiseman** [comment]
> I can see the same denials for `functionscdn.azureedge.net`
> I've added an allow rule for this for the function app IP and the backlog of containers have magically appeared in the storage account.
> Wasn't entirely sure how to find the IP for the function app, but noticed that all the firewall denials were for a specific IP in the Airlock Processor Subnet, so used that.

Can I ask for noddy hints on how to find the errors in a log file?    Same problem but I can't find the right log

**Comment by Danny-Cooke-cCK** [comment]
What TRE version does this issue exist in please ?

**Comment by martinpeck** [comment]
Current release, but I believe it may be an issue for all releases. I suspect this has always been an issue, but it may only present itself in certain situations.

My guess is that if function app is deployed before the firewall then will successfully make this call on start up and then be fine until such time as the function app is redeployed.

In the deployment where I saw this the function app (and therefore the airlock) we working fine and then stopped working.

**Comment by Danny-Cooke-cCK** [comment]
ok thanks. i ask as im seeing the same same symptoms with airlock but just with a normal airlock request. the SA is present but the Container isn't

**Comment by martinpeck** [comment]
Sounds like this issue. You can confirm this by looking at service bus queues (there will be messages waiting to be processed).

**Comment by Danny-Cooke-cCK** [comment]
Thanks @martinpeck the workaround suggested worked for us too. Strange as this was running for months with no issues

**Comment by bwiseman** [comment]
Is there extra logging I need to enable to see the airlock processor firewall decisions?  I have 2 TRE installs

-  one working.  Full redeployment with the malware scan disabled, sleep timer added, airlock working.  There are no log entries in the firewall when I use the airlock.  I'd like to identify the ip and make the firewall change.  
- one broken,  Unable to upload files to the blob container.   No sign of firewall log entries .  Can't seem to identify the functionapp ip address to amend the firewall rules

Terrified now that if I amend anything and 'make all' I'm going to break the airlock again.   I'm also trying to find the service bus queues to check for waiting jobs.... which queue?

**Comment by martinpeck** [comment]
@bwiseman : for the service bus queue, see the original text for this issue.

**Comment by bwiseman** [comment]
Things queued in airlock-status-changed service bus

Nothing at all that I can find anywhere in the TRE firewall logs about `functionscdn.azureedge.net` to identify the functionapp IP

**Comment by marrobi** [comment]
@bwiseman the PR here - https://github.com/microsoft/AzureTRE/pull/3682 - should fix the issue - you could try pulling that branch into your repo and upgrading the firewall.
